modules/javafx.controls/src/main/java/javafx/scene/control/skin/TableViewSkinBase.java

```diff
@@ -508,7 +508,11 @@ protected TableHeaderRow createTableHeaderRow() {
      *                                                                         *
      **************************************************************************/
 
-    final TableHeaderRow getTableHeaderRow() {
+    /**
+     *
+     * @return The TableHeaderRow created using {@link #createTableHeaderRow()}.
+     */
+    protected final TableHeaderRow getTableHeaderRow() {
```

Comment:
If there is a use case to override then it could be made non-final. Is the "override to narrow the type" something that you see as being likely?

Reply:
yes, I do it whenever the custom class created by the factory method has additional public api (vs. just a bug hacking class). My pattern is something like:

// factory method
protected MyType createType() {
    return new MyType();
}

protected MyType getType() {
    // cast is safe because the factory guarantees at least MyType
    return (MyType) super.getType();
}   

That was fine for fx8 (f.i.) TableHeaderRow TableViewSkin, not possible since fx9. The ugly way around a final getter is to either cast at each usage of any extended api or introduce a custom getter with another name, like

protected MyType getMyType() {
    return (MyType) getType();
}

Note that the pattern of overidable factory methods and final getters is spread all over the skin classes, very nasty, IMO.